### ISSUE TITLE
Optimize OpenBLAS for POWER8 VSX

### ISSUE THREAD

**Issue by edelsohn** [xianyi/OpenBLAS#774]
Extend the current PowerPC with VMX support to PPC64 Linux Big Endian and Little Endian, tuned for POWER8 with VSX double precision floating point enablement.

**Comment by brada4** [comment]
Compilers you claim in your profile you write may or may not emit particular instructions.
OpenBLAS is tuned for real processors, your request is approximately - can you optimize it for x86 ISA?

**Comment by xianyi** [comment]
@brada4 , thank you for the comments.

@edelsohn may create some bountysource.com bounty for these issues. We would like to optimize OpenBLAS on other CPU and architectures, including IBM POWER.

**Comment by xianyi** [comment]
I am going to close this issue. I think @wernsaar already finished the optimization for BLAS3 double precision.

**Comment by aszalucha** [comment]
@wernsaar thank you for your contribution, however it's still seems to work slower than standard BLAS/ATLAS. 

For example [this script](https://gist.github.com/alimuldal/eb0f4eea8af331b2a890) gives me with OpenBLAS:
```
version: 1.10.4
maxint:  9223372036854775807

/home/aszalucha/env/lib/python2.7/site-packages/numpy/distutils/system_info.py:635: UserWarning: Specified path  is invalid.
  warnings.warn('Specified path %s is invalid.' % d)
BLAS info:
 * libraries ['openblas', 'openblas']
 * library_dirs ['/opt/OpenBLAS/lib']
 * define_macros [('HAVE_CBLAS', None)]
 * language c

dot: 0.101142 sec

```
while with RHEL's BLAS

```

version: 1.10.4
maxint:  9223372036854775807

/usr/lib64/python2.7/site-packages/numpy/distutils/system_info.py:635: UserWarning: Specified path  is invalid.
  warnings.warn('Specified path %s is invalid.' % d)
BLAS info:
 * libraries ['openblas', 'openblas']
 * library_dirs ['/opt/OpenBLAS/lib']
 * define_macros [('HAVE_CBLAS', None)]
 * language c

dot: 0.020129 sec


```

PS I've made some mess to normal numpy, so it's partially linked to OpenBLAS, and partialy to ATLAS 

```
ldd /usr/lib64/python2.7/site-packages/numpy/core/_dotblas.so
        linux-vdso64.so.1 =>  (0x00003fff96a60000)
        libtatlas.so.3 => /usr/lib64/atlas/libtatlas.so.3 (0x00003fff95960000)
        libpython2.7.so.1.0 => /lib64/libpython2.7.so.1.0 (0x00003fff956a0000)
        libpthread.so.0 => /lib64/power8/libpthread.so.0 (0x00003fff95660000)
        libc.so.6 => /lib64/power8/libc.so.6 (0x00003fff95450000)
        libgfortran.so.3 => /lib64/libgfortran.so.3 (0x00003fff951f0000)
        libm.so.6 => /lib64/power8/libm.so.6 (0x00003fff95100000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00003fff950d0000)
        libutil.so.1 => /lib64/libutil.so.1 (0x00003fff950a0000)
        /lib64/ld64.so.1 (0x000000002a900000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00003fff95060000)


```
And [this](https://gist.githubusercontent.com/osdf/3842524/raw/b021ccf84e5a119bf3ab5225081f7afabb64a0c3/test_scipy.py) gives me with OpenBLAS:
```

cholesky: 0.164514017105 sec
svd: 4.42001838684 sec
```

while with RHEL's BLAS:

```
cholesky: 0.0200778007507 sec
svd: 0.758527994156 sec
```

**Comment by edelsohn** [comment]
@aszalucha This initial issue only implemented BLAS 3 DGEMM.  This issue really should have been a meta-issue and separate issues created for each component. They're separately working on BLAS 3 SGEMM, and BLAS 1/2 double and single precision.

Are your tests of cholesky and svd invoking the one kernel implemented so far?

**Comment by aszalucha** [comment]
OK, it is related to number of cores 

```
(env) [aszalucha@redhat Downloads]$ OMP_NUM_THREADS=1 python test_scipy.py
cholesky: 0.022839641571 sec
svd: 0.968331813812 sec
(env) [aszalucha@redhat Downloads]$ OMP_NUM_THREADS=2 python test_scipy.py
cholesky: 0.0183848381042 sec
svd: 0.786104631424 sec
(env) [aszalucha@redhat Downloads]$ OMP_NUM_THREADS=4 python test_scipy.py
cholesky: 0.0132524013519 sec
svd: 0.789405584335 sec
(env) [aszalucha@redhat Downloads]$ OMP_NUM_THREADS=8 python test_scipy.py
cholesky: 0.0172588348389 sec
svd: 0.773831033707 sec
(env) [aszalucha@redhat Downloads]$ OMP_NUM_THREADS=16 python test_scipy.py
cholesky: 0.0265274047852 sec
svd: 1.10146837234 sec
(env) [aszalucha@redhat Downloads]$ OMP_NUM_THREADS=32 python test_scipy.py
cholesky: 0.038130569458 sec
svd: 2.04456062317 sec
(env) [aszalucha@redhat Downloads]$ OMP_NUM_THREADS=64 python test_scipy.py
cholesky: 0.141304588318 sec
svd: 4.2699051857 sec

```

**Comment by xianyi** [comment]
@aszalucha , it looks like you achieved the best performance on four threads.

**Comment by brada4** [comment]
numpy.random.random is single precision. It tests C kernels and with 2MB sized data sample multiprocessing calibration could wish better, on a PC too.
If you wish double prcision pick python examples from benchmarks in this very github repository (or use R which is double-precision only)

**Comment by edelsohn** [comment]
I suspect that it tested the old ATLAS BLAS VMX code versus the old OpenBLAS VMX code.  The single precision OpenBLAS support is in progress, and ATLAS BLAS support for POWER8 VSX is available in 3.11.38 but the Linux distros have resisted creating a package for POWER based on the 3.11 development branch instead of the 3.10 release branch.

**Comment by brada4** [comment]
RHEL7 atlas is built with ppc64le-remove-vsx.patch, most likely in both cases VMX is not disturbed.
